spec/controllers/ops_controller/settings/schedules_spec.rb

```diff
@@ -243,4 +243,42 @@
       expect(filtered_list.first).to include(ems_infra_with_hosts.name)
     end
   end
+
+  describe ".schedule_run_now" do
+    let(:user) { FactoryBot.create(:user_admin) }
+
+    before do
+      EvmSpecHelper.local_miq_server
+      login_as(user, :stub_controller => true)
+      allow(controller).to receive(:get_node_info)
+      allow(controller).to receive(:replace_right_cell)
+    end
+
+    let(:schedule) { FactoryBot.create(:miq_schedule) }
+
+    it 'adds 1 Schedule' do
+      controller.params["check_#{schedule.id}"] = "1"
+      controller.send(:schedule_run_now)
+      expect(assigns(:flash_array)).to eq([{:message => "The selected Schedule has been queued to run", :level => :success}])
+      expect_miq_queue_created_by(schedule)
+    end
+
+    let(:schedules) { FactoryBot.create_list(:miq_schedule, 10) }
+
+    it 'adds all Schedules' do
+      schedules.each { |schedule| controller.params["check_#{schedule.id}"] = "1" }
+      controller.send(:schedule_run_now)
+      expect(assigns(:flash_array)).to eq([{:message => "The selected Schedules have been queued to run", :level => :success}])
+      schedules.each { |schedule| expect_miq_queue_created_by(schedule) }
+    end
+
+    def expect_miq_queue_created_by(schedule)
```

Comment:
:fire: :fire: :fire: :fire_engine: if you'd really like to share code, please use a shared example

Reply:
what is wrong with defined methods in specs ?

Reply:
Well, you can't really determine in what context they're running and it can mess up other tests. For example you can get into a situation where you run an isolated spec and it passes and when you run a set of specs, you override something that should not be there and the test fails.

There's an exception for helpers that are being defined outside the context of a spec and included from a module. But it's better to avoid this thing because you're not really encapsulating your methods into classes or modules.

Reply:
thanks for explanation,


>... you override something that should not be there and the test fails.

and this situation could happen only when you will accidentally define method with name of method which is already defined in current rspec file ? It seems that it could happen very unlikely 

Can you refer some docs that it is wrong and bad habit ? Because I saw that we are using it normally  - so we need to stop doing it in whole app.

To our case: what is your suggestion? Don't share code, put it into helper or use shared example ?


thanks

Reply:
@lpichler there's no high level documentation of RSpec about how to write tests. You are allowed to do whatever you want and it's always your fault if your tests are failing, i.e. RSpec is not an enterprise tool that you are using for testing. The same goes for Ruby, it's not documented anywhere in the Ruby docs about how to write good code. I think @vandrlov has our book about writing good tests in RSpec btw, I'm pretty sure it mentions how this can be dangerous.

> and this situation could happen only when you will accidentally define method with name of method which is already defined in current rspec file

As the context/describe world is a little different, it's not always isolating the method definition. Therefore, as I explained above, you don't know on what class/module your method is being defined. So it can cause trouble if you have the method defined in multiple files even in multiple repos if they're loaded together (like we do with MiQ). In some cases even a single definition of a single class in a spec can cause trouble, I can show you an example next week at the office.

> To our case: what is your suggestion? Don't share code, put it into helper or use shared example ?

shared examples, as I wrote above